Modules/_sqlite/connection.c

```diff
@@ -2091,7 +2091,14 @@ serialize_impl(pysqlite_Connection *self, const char *name)
                      name);
         return NULL;
     }
-    PyObject *res = PyBytes_FromStringAndSize(data, size);
+#if PY_SSIZE_T_MAX < 9223372036854775807
```

Comment:
Why this check is necessary ?

`PyBytes_FromStringAndSize` already checks for overflow here:

https://github.com/python/cpython/blob/af5bb1fba49c1f703890e28f69f0928109ecd815/Objects/bytesobject.c#L93-L97

Reply:
Using only 32bit version of `PyBytes_FromStringAndSize()` would be ok if SQLite guarantees the data size to be less than `sizeof(int)` with no paging algorithm, or truncated db is always safe and valid.

Reply:
Thanks for pointing that out, Kumar. As you correctly state, there is no need to duplicate that check.

Reply:
FTR, `sqlite3_serialize` outputs the size as `sqlite_int64`.

Reply:
Unfortunately, `PyBytes_FromStringAndSize()` won't work expectedly:
```
#include <Python.h>

#define PyBytesObject_SIZE (offsetof(PyBytesObject, ob_sval) + 1)

int main(int argc, char *argv[])
{
    int64_t large = (int64_t)1<<30;

    Py_ssize_t size = (Py_ssize_t)large;
    for (int i = 1; i < 5; i++) {
        if ((size_t)size > (size_t)PY_SSIZE_T_MAX - PyBytesObject_SIZE) {
            printf("%d: overflow: %u\n", i, size);
        }
        else {
            printf("%d: ok      : %u\n", i, size);
        }
        size <<= 1;
    }
}
```
```
1: ok      : 1073741824
2: overflow: 2147483648
3: ok      : 0
4: ok      : 0
```
Leaving the warning would be better than hiding without information about SQLite.

Reply:
A one-line alternative without `#ifdef`, if acceptable for 64bit platforms:

`size = Py_MIN((size_t)size, (size_t)PY_SSIZE_T_MAX);`

Reply:
Correcting:
`size = Py_MIN((uint64_t)size, (uint64_t)PY_SSIZE_T_MAX);`

Reply:
Let's take a step back and look at what happens on a 32-bit system, and see if we can spot the correct fix.

### Data types
- [Py_ssize_t](https://docs.python.org/3.12/c-api/intro.html#c.Py_ssize_t) is 32 bits
- [int](https://en.wikipedia.org/wiki/C_data_types#Main_types) is at least 16 bits
- `sqlite3_int64` is 64 bits

### Inside serialize_impl()

1. [sqlite3_serialize](https://sqlite.org/c3ref/serialize.html) is called and returns `size` as a 64-bit signed integer (`sqlite3_int64`)
2. [PyBytes_FromStringAndSize](https://docs.python.org/3.12/c-api/bytes.html#c.PyBytes_FromStringAndSize) is called, and it takes `Py_ssize_t` as an argument. `size` is 64 bits, and `Py_ssize_t` is 32 bits. **This is the problem.**


### Proposed solutions so far

- ❌: cast `size` to `Py_ssize_t` (original solution in this PR); potential truncation
- ✅: add overflow check before `PyBytes_FromStringAndSize` (second solution by neonene); this is the correct solution
- ❌: rely on check inside `PyBytes_FromStringAndSize`; nope, `size` has already been cast; potential truncation
- ❌: `Py_MIN(size, PY_SSIZE_T_MAX)`; potential truncation


### The overflow check

- Guarding the overflow check with a preprocessor condition is good; we don't need this check for 64-bit archs. This needs to be in the PR.
- What value should we check for? A possible 16-bit value (`INT_MAX`) is suggested? I don't understand why; it makes no sense for me. The only sensible thing to do, is to check against the _maximum value accepted by the datatype of the `len` argument of `PyBytes_FromStringAndSize`_. That argument is of datatype `Py_ssize_t` and its maximum value is `PY_SSIZE_T_MAX`. This is the value we need to check for, and this needs to be in the PR.


AFAICS, the PR, as it was on commit 3fbf0d02bf85ed5221c756dc5c10886c4d51faa6, is correct. @neonene, can you please update it?

Reply:
I have a follow up question, since `sqlite` has contiguous memory representation of the data, how can it overflow `PY_SSIZE_T_MAX` which on a 32 bit system would be 32 bits. If this were to overflow then that means sqlite can allocate more than `2 ** 32` bytes of memory on 32 bit system, which cannot be true as the virtual address space is limited to 4 GB of memory.

Reply:
That's a good question, Kumar :) I have been wondering this myself. Is this actually possible at all, or is this overflow purely theoretical? MMU's and address space is not my field of expertise, so I need to look this up, but what you say makes sense to me.

Reply:
> [...] since sqlite has contiguous memory representation of the data [...]

Oh, this is not given. It may be fragmented.

Reply:
Even considering PAE (Physical Address Extension), one process is still limited to 4GB on 32 bit platform so this cannot be true even theoretically.

Reply:
The explanation is simple regardless of sqlite's internal representation. `sqlite3_serialize` returns a pointer to an array of bytes and python creates a bytes object from it. On 32 bit platform, even theoretically you cannot return a pointer to an array longer than 4GB in memory so this is not possible.

Reply:
This implies that truncation is purely theoretical, and that a simple cast is enough. I might borrow some of your words in a comment about it.

Thanks for the explanations, Kumar! Highly appreciated.